### PR TITLE
Construction

### PR DESCRIPTION
**Pull request by dannyrhubarb**

### CODE REVIEW

File: src/main/java/Province.java [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+import java.util.Map;
+
+public class Province {
+
+    private final Land land;
+    private final Resources resources;
+
+    public Province(int initialBarrenLand) {
+        land = new Land(initialBarrenLand);
+        resources = new Resources();
```
Comment (dannyrhubarb) [inline src/main/java/Province.java:13]: Prepared for resources. Should resources take Land as dependency?

File: src/main/java/Province.java [hunk 2]
```diff
@@ -0,0 +1,144 @@
+import java.util.AbstractMap;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+public class Province {
+
+    private static final int BUILDING_CONSTRUCTION_TIME = 24;
+    private static final int BUILDING_DESTRUCTION_TIME = 12;
```
Comment (phiwer) [inline src/main/java/Province.java]: This is instant

File: src/test/java/TestProvince.java [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+
+public class TestProvince {
+
+    @Test
+    void testConstruction() {
+        int initialBarrenLand = 100;
+
+        Province province = new Province(initialBarrenLand);
+
+        assertEquals(initialBarrenLand, province.getUndevelopedLand());
```
Comment (phiwer) [inline src/test/java/TestProvince.java:15]: SHould this also be called barrenLand?

File: src/test/java/TestProvince.java [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+                new Building.Request(Building.BANKS, 10),
+                new Building.Request(Building.STABLES, 10),
+                new Building.Request(Building.TOWERS, 10)
+        ));
+
+        int buildingsUnderConstruction = 25+25+10+10+10;
+
+        assertEquals(initialBarrenLand, province.getTotalLand());
+        assertEquals(initialBarrenLand - buildingsUnderConstruction, province.getUndevelopedLand());
+        assertEquals(buildingsUnderConstruction, province.getBuildingsInProgress());
```
Comment (phiwer) [inline src/test/java/TestProvince.java:29]: buildingsInProgress or other way around?

File: src/test/java/TestProvince.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+    @Test
+    void testDestruction() {
+        int initialBarrenLand = 100;
+
+        Province province = new Province(initialBarrenLand);
+
+        int numberOfBuildings = 25;
+
+        province.build(List.of(
+                new Building.Request(Building.FARMS, numberOfBuildings)
```
Comment (phiwer) [inline src/test/java/TestProvince.java:42]: I really like this! I will steal this way of constructing a building request as I am not satisified with mine :)

File: src/main/java/Building.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+    }
+
+    public Optional<Integer> resourceCap() {
+        return switch (this) {
+            case STABLES -> Optional.of(80);
+            default -> Optional.empty();
+        };
+    }
+
+    public static final class Request {
```
Comment (phiwer) [inline src/main/java/Building.java:27]: record?

File: src/test/java/TestProvince.java (hunk 3 above)
Comment (dannyrhubarb) [reply src/test/java/TestProvince.java:15]: Yep, probably. Been experimenting with the terminology

File: src/main/java/Building.java (hunk 6 above)
Comment (dannyrhubarb) [reply src/main/java/Building.java:27]: Yes! I have Java 14 by default so I have to upgrade though

File: src/test/java/TestProvince.java (hunk 4 above)
Comment (dannyrhubarb) [reply src/test/java/TestProvince.java:29]: You mean choose one or the other? This is the result of refactoring to try different terminologies

File: src/main/java/Province.java (hunk 2 above)
Comment (dannyrhubarb) [reply src/main/java/Province.java]: You managed to comment on an old commit here?

File: src/main/java/Province.java (hunk 1 above)
Comment (phiwer) [reply src/main/java/Province.java:10]: The income formula is quite complicated though:

Raw Income = (3 * Employed Peasants) + (1 * Unemployed Peasants) + (0.75 * Prisoners) + (Banks * 25 * BE)

Modified Income = Raw Income * Plague * Riots * Bank % Bonus * Income Sci * Honor Income Mod 
                  * Race Mod * Personality Mod * Dragon * Ritual

Expense formula is as well:

Military Expenses = (((Def specs + Off specs )*0.5) + Elites * 0.75) * Wage Rate * Armouries Bonus * Race Mod * Personality Mod * max(Inspire Army , Hero's Inspiration) * Greed * Bookkeeping Science Effect

**Review by phiwer — CHANGES_REQUESTED** [review]
Please have a look at my formulas wrt resources.

File: src/main/java/Province.java (hunk 1 above)
Comment (dannyrhubarb) [reply src/main/java/Province.java:10]: Great, thanks!

**Pull request merged**